Teleport Router tests ✨

EvolveArt opened the pull request:
EnumerableSet implementation taken from an [OpenZeppelin PR](https://github.com/OpenZeppelin/cairo-contracts/pull/247). Can be improved from there and adapted, for instance index starts at 1 in this implementation compared to the solidity one which starts at 0.

EvolveArt committed:
feat: teleport router tests ✨

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -100,14 +101,16 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
 
     // wards[msg.sender] = 1;
-    _wards.write(caller, 1);
+    _wards.write(ward, 1);
```
Idea here was that contract deployments need to go through an account abstraction. Not sure if it is true.

EvolveArt replied:
I guess we'll see with Cairo 0.10.1 and this proposal aswell https://community.starknet.io/t/universal-deployer-contract-proposal/1864

maciejka replied:
Right, it is safer to use constructor args right now.

EvolveArt committed:
feat: solidity comments + new changes ✨

- Update code to follow https://github.com/makerdao/dss-teleport/blob/routing-v2/src/TeleportRouter.sol
- add solidity comments for easier review

EvolveArt committed:
fix: ci ♻️

EvolveArt committed:
fix: ci ♻️

EvolveArt committed:
Update pyproject.toml

EvolveArt committed:
fix: ci ♻️

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -102,13 +132,18 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
 // constructor(address dai_) {
```
Is solidity version up to date?

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -50,15 +55,33 @@ func _wards(address: felt) -> (res: felt) {
 func _allDomains() -> (res: felt) {
 }
 
+// The minimum amount of DAI to be flushed per target domain (prevent spam)
```
Where is solidity original?

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -50,15 +55,33 @@ func _wards(address: felt) -> (res: felt) {
+func _domain() -> (res: felt) {
+}
+
 // address public parent;
```
Is it up to date?

maciejka replied:
I mean it should be parentDomain.

maciejka replied:
`parentDomain` should be in the Solidity comment

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -76,16 +99,23 @@ func Rely(usr: felt) {
 func File_gateway(what: felt, domain: felt, data: felt) {
 }
 
+@event
```
Solidity version?

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -76,16 +99,23 @@ func Rely(usr: felt) {
+func InitiateTeleport(teleport: TeleportGUID) {
+}
+
+@event
```
Solidity version?

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -102,13 +132,18 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
     let (caller) = get_caller_address();
 
+    // domain = domain_;
+    //     parentDomain = parentDomain_;
```
wrong indentation

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -269,24 +324,41 @@ func gateways{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(d
     }
 
-    let (gateway) = get_gateway(teleportGUID.target_domain, parent);
+    _registerMint(teleportGUID);
```
Solidity seems to be missing.

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -318,23 +378,27 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
 
-    // address gateway = gateways[targetDomain];
-    let (gateway) = _gateways.read(target_domain);
+    _settle(caller, source_domain, target_domain, amount);
```
Solidity seems to be missing.

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -343,14 +407,97 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+}
+
+@external
+func initiateTeleport{
```
Solidity?

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -343,14 +407,97 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+// * @dev Will initiate a settle operation along the secure, slow routing path
+// * @param targetDomain The target domain to settle
+@external
+func flush{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(target_domain: felt) {
```
Solidity

maciejka commented on test/teleport_router.ts:
```diff
@@ -1,33 +1,16 @@
 import { Account } from '@shardlabs/starknet-hardhat-plugin/dist/src/account';
```
In general I miss original test declarations. Keeping them will make alignment easier.

maciejka replied:
Another general remark: indentation of solidity code needs to be fixed.

maciejka replied:
Test declarations are still missing. I mean declarations like: `function testFileRemoveNotLastDomain() public {`. I don't want to need to guess coresponding test names (even if it is an easy guess).

maciejka reviewed (commented):
It is hard to review without complete Solidity in the comments.

EvolveArt committed:
fix: solidity comments ✍️

EvolveArt commented:
> It is hard to review without complete Solidity in the comments.

Fixed in last commits, note that I've moved `initiateTeleport` and `flush` back to `TeleportRouter` following latest changes made [here](https://github.com/makerdao/dss-teleport/blob/routing-v2/src/TeleportRouter.sol)

EvolveArt committed:
fix: typo ✍️

EvolveArt committed:
fix: test comments indentation ✍️

EvolveArt committed:
Update hardhat.config.ts

EvolveArt committed:
fix: switch to docker env ♻️

maciejka commented on contracts/starknet/teleport_router.cairo:
```diff
@@ -98,17 +135,24 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
 
+    let (caller) = get_caller_address();
     // emit Rely(msg.sender);
     Rely.emit(caller);
```
It should be `Rely.emit(ward);`

EvolveArt committed:
fix: review ♻️

EvolveArt committed:
fix: test names comments ✍️

EvolveArt committed:
Update chain-l2.sh

EvolveArt committed:
Merge branch 'domain' into evolve/teleport-router
